File: app/src/main/java/com/example/inappcamerademo/MainDbHelper.java
package com.example.inappcamerademo;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import com.example.inappcamerademo.model.ImageInfo;

import java.util.ArrayList;
import java.util.List;

public class MainDbHelper extends SQLiteOpenHelper {

    private static final String TAG = "MainDbHelper";

    private static final String DATABASE_NAME = "ImageInfoDb";
    private static final int DATABASE_VERSION = 1;

    private static MainDbHelper mInstance;

    private static final String TABLE_NAME = "IMAGE_DATA";
    private static final String KEY_IMAGE_NAME = "NAME";
    private static final String KEY_IMAGE_LAT = "LAT";
    private static final String KEY_IMAGE_LON = "LON";
    private static final String KEY_IMAGE_TIME = "TIME";
    private static final String KEY_IMAGE_PATH = "PATH";


    private static final String CREATE_CONTACTS_TABLE = "CREATE TABLE " + TABLE_NAME + "("
            + KEY_IMAGE_NAME + " TEXT PRIMARY KEY,"
            + KEY_IMAGE_LAT + " DOUBLE,"
            + KEY_IMAGE_LON + " DOUBLE,"
            + KEY_IMAGE_TIME + " DOUBLE,"
            + KEY_IMAGE_PATH + " TEXT"
            + ")";


    private MainDbHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
    }

    public static MainDbHelper getInstance(Context context) {
        if (mInstance == null)
            mInstance = new MainDbHelper(context);
        return mInstance;
    }

    @Override
    public void onCreate(SQLiteDatabase sqLiteDatabase) {
        sqLiteDatabase.execSQL(CREATE_CONTACTS_TABLE);
    }

    @Override
    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i1) {
        sqLiteDatabase.execSQL("DROP TABLE IF EXISTS " + TABLE_NAME);
        onCreate(sqLiteDatabase);
    }

    public void addImage(ImageInfo imageInfo) {
        Log.v(TAG, "addImage:" + imageInfo);

        SQLiteDatabase db = this.getWritableDatabase();

        ContentValues values = new ContentValues();
        values.put(KEY_IMAGE_NAME, imageInfo.getName());
        values.put(KEY_IMAGE_TIME, imageInfo.getTimestamp());
        values.put(KEY_IMAGE_LAT, imageInfo.getLat());
        values.put(KEY_IMAGE_LON, imageInfo.getLon());
        values.put(KEY_IMAGE_PATH, imageInfo.getPath());

        long id = db.insert(TABLE_NAME, null, values);
        Log.v(TAG, "id:" + id);
        db.close();
    }

    public ImageInfo getImageInfoFromImageName(String name) {
        Log.v(TAG, "name:" + name);

        SQLiteDatabase db = this.getReadableDatabase();

        String query = "SELECT * FROM " + TABLE_NAME + " WHERE " + KEY_IMAGE_NAME + " =?";
        Cursor cursor = db.rawQuery(query, new String[]{name});
        ImageInfo imageInfo = null;
        if (cursor != null && cursor.moveToFirst()) {
            imageInfo = new ImageInfo();
            imageInfo.setName(cursor.getString(cursor.getColumnIndex(KEY_IMAGE_NAME)));
            imageInfo.setTimestamp(cursor.getLong(cursor.getColumnIndex(KEY_IMAGE_TIME)));
            imageInfo.setLat(cursor.getDouble(cursor.getColumnIndex(KEY_IMAGE_LAT)));
            imageInfo.setLon(cursor.getDouble(cursor.getColumnIndex(KEY_IMAGE_LON)));
            imageInfo.setPath(cursor.getString(cursor.getColumnIndex(KEY_IMAGE_PATH)));
            cursor.close();
        }
        db.close();
        return imageInfo;
    }

    public List<String> getAllImageNames() {
        SQLiteDatabase db = this.getReadableDatabase();

        String query = "SELECT " + KEY_IMAGE_NAME + " FROM " + TABLE_NAME;
        Cursor cursor = db.rawQuery(query, null);
        List<String> list = new ArrayList<>();

        if (cursor.moveToFirst()) {
            do {
                list.add(cursor.getString(0));
            } while (cursor.moveToNext());
        }

        cursor.close();
        db.close();

        return list;
    }

    public List<String> getAllImageNamesSortedByName() {
        SQLiteDatabase db = this.getReadableDatabase();

        String query = "SELECT " + KEY_IMAGE_NAME + " FROM " + TABLE_NAME + " ORDER BY " + KEY_IMAGE_NAME + " ASC ";
        Cursor cursor = db.rawQuery(query, null);
        List<String> list = new ArrayList<>();

        if (cursor.moveToFirst()) {
            do {
                list.add(cursor.getString(0));
            } while (cursor.moveToNext());
        }

        cursor.close();
        db.close();

        return list;
    }

    public List<String> getAllImageNamesSortedByTime() {
        SQLiteDatabase db = this.getReadableDatabase();

        String query = "SELECT " + KEY_IMAGE_NAME + " FROM " + TABLE_NAME + " ORDER BY " + KEY_IMAGE_TIME + " DESC ";
        Cursor cursor = db.rawQuery(query, null);
        List<String> list = new ArrayList<>();

        if (cursor.moveToFirst()) {
            do {
                list.add(cursor.getString(0));
            } while (cursor.moveToNext());
        }

        cursor.close();
        db.close();

        return list;
    }

    public ImageInfo getLastCapturedImage() {
        SQLiteDatabase db = this.getReadableDatabase();

        String query = "SELECT * FROM " + TABLE_NAME + " ORDER BY " + KEY_IMAGE_TIME + " DESC ";
        Cursor cursor = db.rawQuery(query, null);

        ImageInfo imageInfo = null;
        if (cursor.moveToFirst()) {
            imageInfo = new ImageInfo();
            imageInfo.setName(cursor.getString(cursor.getColumnIndex(KEY_IMAGE_NAME)));
            imageInfo.setTimestamp(cursor.getLong(cursor.getColumnIndex(KEY_IMAGE_TIME)));
            imageInfo.setLat(cursor.getDouble(cursor.getColumnIndex(KEY_IMAGE_LAT)));
            imageInfo.setLon(cursor.getDouble(cursor.getColumnIndex(KEY_IMAGE_LON)));
            imageInfo.setPath(cursor.getString(cursor.getColumnIndex(KEY_IMAGE_PATH)));
        }

        cursor.close();
        db.close();

        return imageInfo;
    }

}

File: app/src/main/java/com/example/inappcamerademo/MainDbHelper.kt
package com.example.inappcamerademo

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import com.example.inappcamerademo.model.ImageInfo
import java.util.*

class MainDbHelper private constructor(context: Context) :
    SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {
    override fun onCreate(sqLiteDatabase: SQLiteDatabase) {
        sqLiteDatabase.execSQL(CREATE_CONTACTS_TABLE)
    }

    override fun onUpgrade(sqLiteDatabase: SQLiteDatabase, i: Int, i1: Int) {
        sqLiteDatabase.execSQL("DROP TABLE IF EXISTS " + TABLE_NAME)
        onCreate(sqLiteDatabase)
    }

    fun addImage(imageInfo: ImageInfo) {
        Log.v(TAG, "addImage:$imageInfo")
        val db = this.writableDatabase
        val values = ContentValues()
        values.put(KEY_IMAGE_NAME, imageInfo.name)
        values.put(KEY_IMAGE_TIME, imageInfo.timestamp)
        values.put(KEY_IMAGE_LAT, imageInfo.lat)
        values.put(KEY_IMAGE_LON, imageInfo.lon)
        values.put(KEY_IMAGE_PATH, imageInfo.path)
        val id = db.insert(TABLE_NAME, null, values)
        Log.v(TAG, "id:$id")
        db.close()
    }

    fun getImageInfoFromImageName(name: String): ImageInfo? {
        Log.v(TAG, "name:$name")
        val db = this.readableDatabase
        val query = "SELECT * FROM " + TABLE_NAME + " WHERE " + KEY_IMAGE_NAME + " =?"
        val cursor = db.rawQuery(query, arrayOf(name))
        var imageInfo: ImageInfo? = null
        if (cursor != null && cursor.moveToFirst()) {
            imageInfo = ImageInfo(
                cursor.getString(cursor.getColumnIndex(KEY_IMAGE_NAME)),
                cursor.getLong(cursor.getColumnIndex(KEY_IMAGE_TIME)),
                cursor.getString(cursor.getColumnIndex(KEY_IMAGE_PATH)),
                cursor.getDouble(cursor.getColumnIndex(KEY_IMAGE_LAT)),
                cursor.getDouble(cursor.getColumnIndex(KEY_IMAGE_LON))
            )
            cursor.close()
        }
        db.close()
        return imageInfo
    }

    val allImageNames: List<String>
        get() {
            val db = this.readableDatabase
            val query = "SELECT " + KEY_IMAGE_NAME + " FROM " + TABLE_NAME
            val cursor = db.rawQuery(query, null)
            val list: MutableList<String> = ArrayList()
            if (cursor.moveToFirst()) {
                do {
                    list.add(cursor.getString(0))
                } while (cursor.moveToNext())
            }
            cursor.close()
            db.close()
            return list
        }
    val allImageNamesSortedByName: List<String>
        get() {
            val db = this.readableDatabase
            val query =
                "SELECT " + KEY_IMAGE_NAME + " FROM " + TABLE_NAME + " ORDER BY " + KEY_IMAGE_NAME + " ASC "
            val cursor = db.rawQuery(query, null)
            val list: MutableList<String> = ArrayList()
            if (cursor.moveToFirst()) {
                do {
                    list.add(cursor.getString(0))
                } while (cursor.moveToNext())
            }
            cursor.close()
            db.close()
            return list
        }
    val allImageNamesSortedByTime: List<String>
        get() {
            val db = this.readableDatabase
            val query =
                "SELECT " + KEY_IMAGE_NAME + " FROM " + TABLE_NAME + " ORDER BY " + KEY_IMAGE_TIME + " DESC "
            val cursor = db.rawQuery(query, null)
            val list: MutableList<String> = ArrayList()
            if (cursor.moveToFirst()) {
                do {
                    list.add(cursor.getString(0))
                } while (cursor.moveToNext())
            }
            cursor.close()
            db.close()
            return list
        }
    val lastCapturedImage: ImageInfo?
        get() {
            val db = this.readableDatabase
            val query = "SELECT * FROM " + TABLE_NAME + " ORDER BY " + KEY_IMAGE_TIME + " DESC "
            val cursor = db.rawQuery(query, null)
            var imageInfo: ImageInfo? = null
            if (cursor.moveToFirst()) {
                imageInfo = ImageInfo(
                    cursor.getString(cursor.getColumnIndex(KEY_IMAGE_NAME)),
                    cursor.getLong(cursor.getColumnIndex(KEY_IMAGE_TIME)),
                    cursor.getString(cursor.getColumnIndex(KEY_IMAGE_PATH)),
                    cursor.getDouble(cursor.getColumnIndex(KEY_IMAGE_LAT)),
                    cursor.getDouble(cursor.getColumnIndex(KEY_IMAGE_LON))
                )
            }
            cursor.close()
            db.close()
            return imageInfo
        }

    companion object {
        private const val TAG = "MainDbHelper"
        private const val DATABASE_NAME = "ImageInfoDb"
        private const val DATABASE_VERSION = 1
        private var mInstance: MainDbHelper? = null
        private const val TABLE_NAME = "IMAGE_DATA"
        private const val KEY_IMAGE_NAME = "NAME"
        private const val KEY_IMAGE_LAT = "LAT"
        private const val KEY_IMAGE_LON = "LON"
        private const val KEY_IMAGE_TIME = "TIME"
        private const val KEY_IMAGE_PATH = "PATH"
        private const val CREATE_CONTACTS_TABLE = ("CREATE TABLE " + TABLE_NAME + "("
                + KEY_IMAGE_NAME + " TEXT PRIMARY KEY,"
                + KEY_IMAGE_LAT + " DOUBLE,"
                + KEY_IMAGE_LON + " DOUBLE,"
                + KEY_IMAGE_TIME + " DOUBLE,"
                + KEY_IMAGE_PATH + " TEXT"
                + ")")

        fun getInstance(context: Context): MainDbHelper? {
            if (mInstance == null) mInstance = MainDbHelper(context)
            return mInstance
        }
    }
}